collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormDownloaderTest.java

```diff
@@ -283,6 +284,59 @@ public void whenFormIsSoftDeleted_unDeletesForm() throws Exception {
         assertThat(formsRepository.get(1L).isDeleted(), is(false));
     }
 
+    @Test
+    public void whenFormWithSameIdAndVersion_butDifferentContentsOnDevice_failsToDownload() throws Exception {
+        Form form = buildForm(1L, "form", "version", getFormFilesPath(), "contents1").build();
+        formsRepository.save(form);
+
+        ServerFormDetails serverFormDetails = new ServerFormDetails(
+                form.getDisplayName(),
+                "http://downloadUrl",
+                null,
+                form.getJrFormId(),
+                form.getJrVersion(),
+                "differenthash",
+                false,
+                true,
+                null);
+
+        FormSource formSource = mock(FormSource.class);
+        when(formSource.fetchForm("http://downloadUrl")).thenReturn(new ByteArrayInputStream("foo".getBytes()));
+
+        ServerFormDownloader downloader = new ServerFormDownloader(formSource, formsRepository, cacheDir, formsDir.getAbsolutePath(), new FormMetadataParser(ReferenceManager.instance()));
+
+        try {
+            downloader.downloadForm(serverFormDetails, null, null);
+            fail("Expected exception");
+        } catch (FormDownloadException e) {
+            assertThat(e.getMessage(), notNullValue());
+        }
+    }
+
+    @Test
+    public void whenFormWithSameIdAndVersion_andSameContentsOnDevice_downloadsSuccessfully() throws Exception {
```

Comment:
Don't want to be too nitpicky about test names but just because this stuff is all kinda confusing I'm wondering if it should be `_replacesForm` or `_doesNotDuplicateForm`?

Reply:
I find naming really important so not nitpicking at all!